virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,208 @@
+CELSIUS_TO_KELVIN = 273.15
+"""Factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+SECOND_TO_DAY = 86400
+"""Factor to convert between days and seconds."""
+
+
+class Radiation:
+    """Radiation balance."""
+
+    def __init__(self, elevation: NDArray[np.float32]) -> None:
+        """Initializes point-based radiation method."""
+
+        self.elevation = elevation
+        """Elevation above sea level, [m]"""
+
+        # populated later
+        self.ppfd = NDArray[np.float32]
+        """Daily top of canopy photosynthetic photon flux density, [mol m-2]"""
+        self.topofcanopy_radiation = NDArray[np.float32]
+        """Daily top of canopy downward shortwave radiation, [J m-2]"""
+        self.longwave_canopy = NDArray[np.float32]
+        """Daily longwave radiation from n individual canopy layers, [J m-2]"""
+        self.longwave_soil = NDArray[np.float32]
+        """Daily longwave radiation from soil, [J m-2]"""
+        self.netradiation_surface = NDArray[np.float32]
+        """Daily net shortwave radiation at the surface (= forest floor), [J m-2]"""
+
+    def calc_ppfd(
+        self,
+        shortwave_in: NDArray[np.float32],
+        sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+        albedo_vis: NDArray[np.float32] = np.array(0.03, dtype=np.float32),
+    ) -> NDArray[np.float32]:
+        """Calculates top of canopy photosynthetic photon flux density [mol m-2].
+
+        Args:
+            shortwave_in: daily downward shortwave radiation [J m-2]
+            sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+                and 1 (cloud free sky)
+            albedo_vis: visible light albedo, default = 0.03
+
+        Returns:
+            tau: atmospheric transmissivity, unitless
+
+        Reference: :cite:t:`Davis2017`
+        """
+        # calculate transmissivity (tau), unitless
+        tau_o = CLOUDY_TRANSMISSIVITY + TRANSMISSIVITY_COEFFICIENT * sunshine_fraction
+        tau = tau_o * (1.0 + BEER_REGRESSION * self.elevation)
+
+        # Calculate daily photosynth. photon flux density (ppfd_d), mol/m^2
+        self.ppfd = (1.0e-6) * FLUX_TO_ENERGY * (1.0 - albedo_vis) * tau * shortwave_in
+        return tau
+
+    def calc_topofcanopy_radiation(
+        self,
+        tau: NDArray[np.float32],
+        shortwave_in: NDArray[np.float32],
+        sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+        albedo_shortwave: NDArray[np.float32] = np.array(0.17, dtype=np.float32),
+    ) -> None:
+        """Calculate top of canopy shortwave radiation [J m-2].
+
+        Args:
+            tau: atmospheric transmissivity, unitless
+            shortwave_in: daily downward shortwave radiation [J m-2]
+            sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+                and 1 (cloud free sky)
+            albedo_shortwave: shortwave albedo, default = 0.17
+        """
```

Comment:
I'd add a sanity-check that `sunshine_fraction` is between 0 and 1.

Reply:
There are also a bunch of other places where similar sanity-checks would be useful.